File: backend/app/services/pricing.py
import json
from typing import Any
from sqlalchemy.orm import Session
from ..models import GlobalSettings


def _get_setting(db: Session, key: str, default: Any) -> Any:
    setting = db.query(GlobalSettings).filter(GlobalSettings.key == key).first()
    if not setting or setting.value is None:
        return default
    return setting.value


def _parse_json(value: Any, fallback: Any) -> Any:
    if value is None:
        return fallback
    try:
        return json.loads(value)
    except (TypeError, ValueError):
        return fallback


def _parse_number(value: Any, fallback: float) -> float:
    try:
        num = float(value)
    except (TypeError, ValueError):
        return fallback
    return num if num == num else fallback


def calculate_price(db: Session, duration_minutes: int, is_vr: bool = False) -> float:
    base_15 = _parse_number(_get_setting(db, "pricing_base_15min", 5), 5)
    base_per_min = _parse_number(_get_setting(db, "pricing_base_per_min", base_15 / 15), base_15 / 15)
    legacy_vr = _parse_number(_get_setting(db, "pricing_vr_surcharge", 0), 0)
    vr_surcharge_per_min = _parse_number(_get_setting(db, "pricing_vr_surcharge_per_min", legacy_vr / 15), legacy_vr / 15)

    rate_list = _parse_json(_get_setting(db, "pricing_duration_rates", "[]"), [])
    rate_map = {}
    if isinstance(rate_list, list):
        for item in rate_list:
            try:
                minutes = int(item.get("minutes"))
                price = float(item.get("price"))
                rate_map[minutes] = price
            except (TypeError, ValueError, AttributeError):
                continue

    total = rate_map.get(duration_minutes, duration_minutes * base_per_min)

    if is_vr and vr_surcharge_per_min > 0:
        total += duration_minutes * vr_surcharge_per_min

    discount_list = _parse_json(_get_setting(db, "pricing_discounts", "[]"), [])
    if isinstance(discount_list, list):
        for item in discount_list:
            try:
                minutes = int(item.get("minutes"))
                if minutes != duration_minutes:
                    continue
                discount_type = item.get("type")
                value = float(item.get("value"))
                if discount_type == "percent":
                    total -= total * (value / 100)
                else:
                    total -= value
                break
            except (TypeError, ValueError, AttributeError):
                continue

    if total < 0:
        total = 0

    return round(total, 2)

File: frontend/src/utils/pricing.ts
type SettingsItem = { key: string; value: string };

export type PricingRate = { minutes: number; price: number };
export type PricingDiscount = { minutes: number; type: 'flat' | 'percent'; value: number };

export type PricingConfig = {
    basePerMin: number;
    vrSurchargePerMin: number;
    rates: PricingRate[];
    discounts: PricingDiscount[];
    allowManualOverride: boolean;
};

const getSettingValue = (settings: SettingsItem[], key: string) =>
    settings.find((item) => item.key === key)?.value;

const parseJsonSetting = <T>(settings: SettingsItem[], key: string, fallback: T): T => {
    const raw = getSettingValue(settings, key);
    if (!raw) return fallback;
    try {
        return JSON.parse(raw) as T;
    } catch {
        return fallback;
    }
};

const parseNumberSetting = (settings: SettingsItem[], key: string, fallback: number) => {
    const raw = getSettingValue(settings, key);
    if (!raw) return fallback;
    const value = Number(raw);
    return Number.isFinite(value) ? value : fallback;
};

export const getPricingConfig = (settings: SettingsItem[]): PricingConfig => {
    const base15 = parseNumberSetting(settings, 'pricing_base_15min', 5);
    const basePerMin = parseNumberSetting(settings, 'pricing_base_per_min', base15 / 15);
    const legacyVr = parseNumberSetting(settings, 'pricing_vr_surcharge', 0);
    const vrSurchargePerMin = parseNumberSetting(settings, 'pricing_vr_surcharge_per_min', legacyVr / 15);
    const rates = parseJsonSetting<PricingRate[]>(settings, 'pricing_duration_rates', [])
        .filter((rate) => Number.isFinite(rate.minutes) && Number.isFinite(rate.price))
        .map((rate) => ({ minutes: Number(rate.minutes), price: Number(rate.price) }))
        .sort((a, b) => a.minutes - b.minutes);
    const discounts = parseJsonSetting<PricingDiscount[]>(settings, 'pricing_discounts', [])
        .filter((rule) => Number.isFinite(rule.minutes) && Number.isFinite(rule.value))
        .map((rule) => ({
            minutes: Number(rule.minutes),
            type: rule.type === 'percent' ? 'percent' : 'flat',
            value: Number(rule.value)
        }))
        .sort((a, b) => a.minutes - b.minutes);
    const allowManualOverride = getSettingValue(settings, 'pricing_allow_manual_override') === 'true';

    return { basePerMin, vrSurchargePerMin, rates, discounts, allowManualOverride };
};

export const calculatePrice = (durationMinutes: number, isVr: boolean, config: PricingConfig) => {
    const rateMap = new Map(config.rates.map((rate) => [rate.minutes, rate.price]));
    let total = rateMap.get(durationMinutes) ?? durationMinutes * config.basePerMin;

    if (isVr && config.vrSurchargePerMin > 0) {
        total += durationMinutes * config.vrSurchargePerMin;
    }

    const discount = config.discounts.find((rule) => rule.minutes === durationMinutes);
    if (discount) {
        if (discount.type === 'percent') {
            total -= total * (discount.value / 100);
        } else {
            total -= discount.value;
        }
    }

    return Math.max(0, Math.round(total * 100) / 100);
};

export const getDurationOptions = (config: PricingConfig, fallback: number[] = [10, 15, 30, 60]) => {
    if (config.rates.length > 0) {
        return config.rates.map((rate) => rate.minutes);
    }
    return fallback;
};
